ops/testing.py

```diff
@@ -3270,6 +3353,41 @@ def get_notices(
         user_id: Optional[int] = None,
         types: Optional[Iterable[Union[pebble.NoticeType, str]]] = None,
         keys: Optional[Iterable[str]] = None,
-        after: Optional[datetime.datetime] = None,
     ) -> List[pebble.Notice]:
-        raise NotImplementedError(self.get_notices)
+        # Similar logic as api_notices.go:v1GetNotices in Pebble.
+
+        filter_user_id = 0  # default is to filter by request UID (root)
+        if user_id is not None:
+            filter_user_id = user_id
+        if select is not None:
+            if user_id is not None:
+                raise self._api_error(400, 'cannot use both "select" and "user_id"')
+            filter_user_id = None
+
+        if types is not None:
+            types = [(t.value if isinstance(t, pebble.NoticeType) else t) for t in types]
+        if keys is not None:
+            keys = list(keys)
+
+        notices: List[pebble.Notice] = []
+        for notice in self._notices.values():
+            if not self._notice_matches(notice, filter_user_id, types, keys):
+                continue
+            notices.append(notice)
+
+        notices.sort(key=lambda notice: notice.last_repeated)
+        return notices
+
+    @staticmethod
+    def _notice_matches(notice: pebble.Notice,
+                        user_id: Optional[int] = None,
+                        types: Optional[List[str]] = None,
+                        keys: Optional[List[str]] = None) -> bool:
+        # Same logic as NoticeFilter.matches in Pebble.
+        if user_id is not None and not (notice.user_id is None or user_id == notice.user_id):
```

Comment:
It could be me, but I had to read this a few times to follow it. I assume that since this comes from the Pebble code a bunch of people already read essentially the same line but in Go and had no trouble with it, so maybe just ignore this.

Perhaps a comment would help? Like:

`# If the notice is not either public or has a matching user ID, the notice doesn't match.`

Reply:
Added a comment as to the logic (for this and the others).